streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -334,10 +335,33 @@ public void close(final Map<TopicPartition, Long> offsets) throws IOException {
 
     @Override
     public void checkpoint(final Map<TopicPartition, Long> offsets) {
+
+        // Find non persistent store's topics
+        final Map<String, String> storeToChangelogTopic = topology.storeToChangelogTopic();
+        final Set<String> globalNonPersistentStoresTopics = new HashSet<>();
+        for (final StateStore store : topology.globalStateStores()) {
+            if (!store.persistent() && storeToChangelogTopic.containsKey(store.name())) {
+                globalNonPersistentStoresTopics.add(storeToChangelogTopic.get(store.name()));
+            }
+        }
+
```

Comment:
Since we can handle the case in the restoration phase above, I think we do not need to use a separate globalNonPersistentStoresTopics here anymore. Instead, we can do the following inside this function:

1. Filter the entry of the pass-in `offsets` map if `!store.persistent() || storeToChangelogTopic.containsKey(store.name())`.

2. checkpointableOffsets.putAll(filteredOffsets);
2.a. In line 245 above, we can still only heck if `checkpoint != null`.

3. if (!filteredOffsets.isEmpty()) filteredOffsets

Note that after the restoration is done, we will fill in the restored offset in line 287:

```
checkpointableOffsets.put(topicPartition, offset);
```

So after the restoration phase we should have the checkpointableOffsets map populated already.

Reply:
```
        final Map<TopicPartition, Long> filteredOffsets = new HashMap<>(offsets);

        for (final StateStore store : topology.globalStateStores()) {
            if (!store.persistent() && storeToChangelogTopic.containsKey(store.name())) {
                String key = storeToChangelogTopic.get(store.name());
                filteredOffsets.remove(????);
            }
        }
```
In order to remove from filtered offsets, I have to iterate all `filteredOffsets` in that place above.

```
// store name -> topic name
Map<String, String> storeToChangelogTopic = topology.storeToChangelogTopic();

// topic name + partition -> long
Map<TopicPartition, Long> offsets
```

We need to map these 2 data structures. That is why I had to use `globalNonPersistentStoresTopics` to reduce iterations from `N X M` to `N + M`. 

Make sense @guozhangwang?

Reply:
I see your point now, that makes sense. However, in line 348 below we should still pass in the filtered offsets into `checkpointableOffsets`, as it will be read in line 246 above. Right?

Reply:
I am not sure. In line 288 `checkpointableOffsets.put(topicPartition, offset);` is used. So filtering only `offsets` can be not true since checkpointableOffsets is changed in line 288 at application starts. So we need to filter both `checkpointableOffsets` and `offsets` together. 

Also `org.apache.kafka.streams.processor.internals.GlobalStateManagerImpl#checkpointed` is public API. Changing `checkpointableOffsets` sounds very dangerous to me. I have handled some cases but I am not sure that I have covered all possibilities. 

`org.apache.kafka.streams.processor.internals.GlobalStateUpdateTask#initialize` for example returns `Map<TopicPartition, Long>` which is actually `stateMgr.checkpointed()`.

What do you think?

Reply:
For global state stores, here is the ordering of each stage:

1) Initialization: `GlobalStreamThread.initialize()` -> `GlobalStateUpdateTask.initialize()` -> `GlobalStateManagerImpl.initialize()`, where we read the checkpoint file into `checkpointableOffsets`.

2) Restoration: In the same `GlobalStateManagerImpl.initialize()`, we call `stateStore.init()`, in which `GlobalStateManagerImpl.register()` is called, and hence `restoreState()` will read from the loaded `checkpointableOffsets`: if it contains offset seekTo(), otherwise seekToBeginning().

3) Starting: The restoration will bootstrap the global stores up to the log end offset, and after that we will write the restored offset to `checkpointableOffsets`: i.e. we will update the map, with the new values. At this stage the non-persistent stores' offsets should be written to it as well (i.e. line 288). Then we will call `GlobalStateUpdateTask.initTopology` to create the update node and go ahead the normal execution. So here the returned `stateMgr.checkpointed()` should already contain the restored offset already, therefore we can safely call `globalConsumer.seek()` in its caller now.

4) Checkpointing: When we call checkpoint(), we should make sure that non-persistent stores are not written to the checkpoint file, and actually whether we should filter on the `checkpointableOffsets` does not affect correctness anyways since we do not use it anywhere anymore, but to be consistent with its name I think it is still better to filter out those non-checkpointing offsets.

Note that the whole logic is a bit awkward as it was spin off the `ProcessorStateManager` class, and as I mentioned above we can consider consolidating them in the future.

Reply:
>I see your point now, that makes sense. However, in line 348 below we should still pass in the filtered offsets into checkpointableOffsets, as it will be read in line 246 above. Right?

I agree with @guozhangwang that `filteredOffsets` should be added to `checkpointableOffsets` or updated with the results of filteredOffsets.  

I say this because it looks to me that during the `restoreState` call on line 245 that a non-persistent global store could end up restoring from a checkpoint when it was determined on line 356 it should not use checkpoints.

Reply:
This call chain: 

org.apache.kafka.streams.processor.internals.GlobalStreamThread#run 
org.apache.kafka.streams.processor.internals.GlobalStreamThread#initialize
org.apache.kafka.streams.processor.internals.GlobalStreamThread.StateConsumer#initialize
org.apache.kafka.clients.consumer.KafkaConsumer#seek

has: 

```
 if (offset < 0)
            throw new IllegalArgumentException("seek offset must not be a negative number");
```

is causing exceptions if we start to store offsets negative in checkpoint file.

Reply:
I offer to store actual offsets in `checkpointableOffsets` but persist as -1 in case a global store to be persisted.

Reply:
@bbejeck I could not get your point since restoring phase is handling negative offsets in line 245. If it is persisted as negative, it means this is a global store and must be consumed from beginning. I am missing something?

Reply:
> is causing exceptions if we start to store offsets negative in checkpointableOffsets.

Hmm.. before `org.apache.kafka.streams.processor.internals.GlobalStreamThread.StateConsumer#initialize` is called, we call

```
GlobalStateUpdateTask#initialize ->  // as stateMaintainer.initialize();
GlobalStateManager#initialize ->
stateStore#init(..) for each store ->
ProcessorContextImpl#register(..) for each store ->
GlobalStateManager#register(..) for each store ->
GlobalStateManager#restoreState
```

At the end of `GlobalStateManager#restoreState` we should have completed restoration (step 2 above) and hence the `checkpointableOffsets` should be fully populated (i.e. all stores will have a non-negative offset value) at the end when we call `stateMgr.checkpointed();` right?